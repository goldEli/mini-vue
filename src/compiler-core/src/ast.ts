export enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ROOT,
  ELEMENT,
}
