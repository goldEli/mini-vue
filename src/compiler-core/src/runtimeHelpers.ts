export const TO_DISPLAY_STRING = Symbol("toDisplayString");
export const CREATE_ELEMENT_BLOCK = Symbol("createElementBlock")

export const helperMapNames = {
  [TO_DISPLAY_STRING]: "toDisplayString",
  [CREATE_ELEMENT_BLOCK]: "createElementBlock"
};
