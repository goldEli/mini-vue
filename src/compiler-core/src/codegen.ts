import { isString } from "../../shared";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  TO_DISPLAY_STRING,
  helperMapNames,
} from "./runtimeHelpers";

const functionName = "render";
export function generate(ast) {
  const context = createContext(ast);

  const args = ["_ctx", "_cache", "$props", "$setup", "$data", "$options"];

  const signature = args.join(", ");
  genFunctionPreamble(ast, context);

  context.push(`export function ${functionName} (${signature}){`);

  context.push(`return `);

  genNode(ast.codegenNode, context);

  context.push(`}`);

  // console.log(context);
  return {
    code: context.code,
  };
}

function genFunctionPreamble(ast, context) {
  const VueBinging = "Vue";

  const helpers = ast.helpers;

  if (helpers.length > 0) {
    const helpersString = ast.helpers
      .map((helper) => `${helper} as _${helper}`)
      ?.join(", ");

    context.push(`import { ${helpersString} } from '${VueBinging}'`);
    context.push(`\n`);
  }
}

function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genSimpleExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPOUND:
      genCompound(node, context);
      break;
    default:
      break;
  }
}

function genCompound(node, context) {
  const { children } = node;
  for (let i = 0; i < children.length; ++i) {
    const child = children[i];

    if (isString(child)) {
      genStr(child, context);
    } else {
      genNode(children[i], context);
    }
  }
}

function genElement(node, context) {
  const { tag, children, props } = node;
  context.push(`${context.helper(CREATE_ELEMENT_VNODE)}(`);
  const str = genNullList([`'${tag}'`, props]).join(", ");
  context.push(str);
  context.push(", ");
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    genNode(child, context);
  }
  context.push(")");
}

const genNullList = (list) => {
  return list.map((item) => item || "null");
};

function genText(node, context) {
  context.push(`'${node.content}'`);
}

function genStr(str, context) {
  context.push(`${str}`);
}

function genSimpleExpression(node, context) {
  context.push(`${node.content}`);
}

function genInterpolation(node, context) {
  context.push(`${context.helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);

  context.push(")");
}

function createContext(ast) {
  const context = {
    ast,
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapNames[key]}`;
    },
  };
  return context;
}
