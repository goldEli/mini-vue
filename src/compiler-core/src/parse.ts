import { NodeTypes } from "./ast";

interface Context {
  source: string;
}

const enum TagType {
  Start,
  End,
}

export function baseParse(source: string) {
  const context = createParserContext(source);
  const root = createRoot(parseChildren(context));
  return root;
}

const interpolationStart = "{{";
const interpolationEnd = "}}";

function parseChildren(context: Context) {
  const nodes: any[] = [];
  const s = context.source;
  let node;
  if (s.startsWith(interpolationStart)) {
    // 处理插值
    node = parseInterpolation(context);
  } else if (s.startsWith("<")) {
    if (/[a-z]/i.test(s[1])) {
      // 处理element
      node = parseElement(context);
    }
  } else {
    node = parseText(context);
  }
  nodes.push(node);

  return nodes;
}
function parseText(context: Context) {
  let lastIndex = context.source.length;
  if (context.source.indexOf(interpolationStart) !== -1) {
    lastIndex = context.source.indexOf(interpolationStart);
  }

  if (context.source.indexOf("<") !== -1) {
    const idx = context.source.indexOf("<");
    lastIndex = lastIndex < idx ? lastIndex : idx;
  }
  const content = parseTextData(context, lastIndex);
  advanceBy(context, content.length);

  return {
    type: NodeTypes.TEXT,
    content,
  };
}

function parseTextData(context: Context, length: number) {
  const content = context.source.slice(0, length);
  advanceBy(context, length);
  return content;
}

function parseElement(context: Context) {
  const element = parseTag(context, TagType.Start);
  parseTag(context, TagType.End);
  return element;
}

function parseTag(context: Context, type: TagType) {
  const { tag, raw } = getTag(context);

  advanceBy(context, raw.length); // 跳过标签名和标签结束符

  if (type === TagType.End) return;

  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

function getTag(context: Context) {
  const reg = /<\/?([a-z]*)>/i;
  const match = context.source.match(reg);
  if (match) {
    return {
      tag: match[1],
      raw: match[0],
    };
  }
  return {
    tag: "",
    raw: "",
  };
}

function createRoot(children) {
  return {
    type: NodeTypes.ROOT,
    children,
  };
}

function parseInterpolation(context: Context) {
  advanceBy(context, interpolationStart.length);
  const closeIndex = context.source.indexOf(interpolationEnd);

  const rawContent = parseTextData(context, closeIndex);
  const content = rawContent.trim();

  advanceBy(context, interpolationEnd.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}

function advanceBy(context: Context, length) {
  context.source = context.source.slice(length);
}
function createParserContext(source: string) {
  return {
    source,
  };
}
