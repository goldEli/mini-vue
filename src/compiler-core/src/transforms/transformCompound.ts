import { NodeTypes } from "../ast";

function isText(node) {
  return node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION;
}

export function transformCompound(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    const children = node.children;

    let compoundNode;
    for (let i = 0; i < children.length; i++) {
      const child = children[i];

      if (isText(child)) {
        compoundNode = {
          type: NodeTypes.COMPOUND,
          children: [child],
        };
        for (let j = i + 1; j < children.length; j++) {
          const next = children[j];
          if (isText(next)) {
            compoundNode.children.push(next);
            children[i] = compoundNode;
            children.splice(j, 1);
            --j;
          } else {
            compoundNode = undefined;
            break;
          }
        }
      }
    }
  }
}
