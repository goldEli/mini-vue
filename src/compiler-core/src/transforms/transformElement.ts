import { NodeTypes } from "../ast";


export function transformElement(node) {
    if (node.type === NodeTypes.ELEMENT) {
        console.log(node)
    }
}