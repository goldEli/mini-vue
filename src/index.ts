export * from "./runtime-dom/index";

import { baseCompile } from "./compiler-core/src";
import { registerCompileFunction } from "./runtime-core/component";
import * as runtimeDom from "./runtime-dom";

function compileToFunction(template) {
  const { code } = baseCompile(template);

  const render = new Function("Vue", code)(runtimeDom);
  return render;
}

registerCompileFunction(compileToFunction);
