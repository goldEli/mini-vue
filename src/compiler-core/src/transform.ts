import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING, helperMapNames } from "./runtimeHelpers";

interface Options {
  nodeTransforms?: any[];
}

interface Context {
  ast: any;
  nodeTransforms: Options["nodeTransforms"];
  helpers: Map<string, 1>;
  helper: (key: string) => void;
}

export function transform(ast, options: Options = {}) {
  const context = createContext(ast, options);
  traverseNode(ast, context);
  createRootCodegen(ast);
  ast.helpers = [...context.helpers.keys()];
}

function createRootCodegen(root) {
  root.codegenNode = root.children[0];
}

function createContext(ast, options: Options = {}) {
  const nodeTransforms = options.nodeTransforms || [];
  const context = {
    ast,
    nodeTransforms,
    helpers: new Map(),
    helper: (key) => {
      context.helpers.set(key, 1);
    },
  };
  return context;
}

function traverseNode(node, context: Context) {
  context?.nodeTransforms?.forEach((transform) => {
    // 遍历插件列表
    transform(node);
  });
  traverseChildren(node, context);
}

function traverseChildren(node, context: Context) {
  const children = node.children || [];

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(helperMapNames[TO_DISPLAY_STRING]);
      break;

    default:
      break;
  }

  for (let i = 0; i < children.length; i++) {
    const node = children[i];
    traverseNode(node, context); // 递归遍历子节点
  }
}
