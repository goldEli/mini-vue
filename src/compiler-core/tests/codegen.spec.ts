import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformElement } from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";


describe('codegen', () => {
    it('string', () => {
       const ast = baseParse("Hello World") 
       transform(ast)

       const {code} = generate(ast)

       expect(code).toMatchSnapshot(); // 生成快照并比较
    });


    it('interpolation', () => {
       const ast = baseParse("{{message}}") 
       transform(ast, {
         nodeTransforms: [transformExpression]
       })
       const {code} = generate(ast)

       expect(code).toMatchSnapshot(); // 生成快照并比较
    });


    it('element', () => {
       const ast = baseParse("<div></div>") 
       transform(ast, {
         nodeTransforms: [transformElement]
       })
       const {code} = generate(ast)

       expect(code).toMatchSnapshot(); // 生成快照并比较
    });
    
});