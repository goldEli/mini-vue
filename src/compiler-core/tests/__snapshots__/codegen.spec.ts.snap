// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"import { createElementBlock as _createElementBlock } from 'Vue'
export function render (_ctx, _cache, $props, $setup, $data, $options){return _createElementBlock('div')}"
`;

exports[`codegen interpolation 1`] = `
"import { toDisplayString as _toDisplayString } from 'Vue'
export function render (_ctx, _cache, $props, $setup, $data, $options){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `"export function render (_ctx, _cache, $props, $setup, $data, $options){return 'Hello World'}"`;
